cmd/snap-install/volmgr/sfdisk.go

```diff
@@ -0,0 +1,236 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+package volmgr
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"os/exec"
+	"strings"
+
+	"github.com/snapcore/snapd/gadget"
+	"github.com/snapcore/snapd/osutil"
+)
+
+var (
+	sectorSize gadget.Size = 512
+)
+
+// sfdiskDeviceDump represents the sfdisk --dump JSON output format.
+type sfdiskDeviceDump struct {
+	PartitionTable sfdiskPartitionTable `json:"partitiontable"`
+}
+
+type sfdiskPartitionTable struct {
+	Label      string            `json:"label"`
+	ID         string            `json:"id"`
+	Device     string            `json:"device"`
+	Unit       string            `json:"unit"`
+	FirstLBA   uint64            `json:"firstlba"`
+	LastLBA    uint64            `json:"lastlba"`
+	Partitions []sfdiskPartition `json:"partitions"`
+}
+
+type sfdiskPartition struct {
+	Node  string `json:"node"`
+	Start uint64 `json:"start"`
+	Size  uint64 `json:"size"`
+	Type  string `json:"type"`
+	UUID  string `json:"uuid"`
+	Name  string `json:"name"`
+}
+
+type SFDisk struct {
+	device         string
+	partitionTable *sfdiskPartitionTable
+}
+
+func newSFDisk(device string) *SFDisk {
+	return &SFDisk{
+		device: device,
+	}
+}
+
+func (sf *SFDisk) deviceInfo() (*gadget.LaidOutVolume, error) {
+	output, err := exec.Command("sfdisk", "--json", "-d", sf.device).CombinedOutput()
+	if err != nil {
+		return nil, osutil.OutputErr(output, err)
+	}
+
+	var dump sfdiskDeviceDump
+	if err := json.Unmarshal(output, &dump); err != nil {
+		return nil, err
+	}
+
+	pv, err := positionedVolumeFromDump(&dump)
+	if err != nil {
+		return nil, err
+	}
+
+	sf.partitionTable = &dump.PartitionTable
+
+	return pv, nil
+}
+
+func (sf *SFDisk) createPartitions(positionedVolume *gadget.LaidOutVolume, usedPartitions []bool, deviceMap map[string]string) error {
+	var buf bytes.Buffer
+
+	// Write partition data in sfdisk dump format
+	ptable := sf.partitionTable
+	buf.WriteString(fmt.Sprintf("label: %s\nlabel-id: %s\ndevice: %s\nunit: %s\nfirst-lba: %d\nlast-lba: %d\n\n",
```

Comment:
I think you can use "fmt.Fprint(buf, "label: %s\n",...)" here which is slightly shorter :). bytes.Buffer implements an io.Writer which is pretty cool.

Reply:
Nice. Fixing as proposed.